test.sh

```diff
@@ -0,0 +1,236 @@
+    if [[ -f "$1" ]]; then
+        return
+    fi
+    fail "$1 does not exist"
+}
+
+function assert_file_eq() {
+    if [[ $(cat "$1") == "$2" ]]; then
+        return
+    fi
+    fail "file $1 does contain '$2': $(cat $1)"
+}
+
+#########################
+# main
+#########################
+REPO="$DIR/repo"
+mkdir "$REPO"
+
+ROOT="$DIR/root"
+function clean_root() {
+    rm -rf "$ROOT"
+    mkdir "$ROOT"
+}
+
+# Init the temp repo.
+TESTCASE=init
+git -C "$REPO" init
+touch "$REPO"/file
+git -C "$REPO" add file
+git -C "$REPO" commit -aqm "init file"
+
+# Build git-sync
+make >/dev/null
+GIT_SYNC=./bin/git-sync-amd64
+
+# Test HEAD one-time
+testcase "head-once"
+echo "$TESTCASE" > "$REPO"/file
+git -C "$REPO" commit -qam "$TESTCASE"
+$GIT_SYNC \
+    --repo="$REPO" \
+    --branch=master \
+    --rev=HEAD \
+    --root="$ROOT" \
+    --dest="link" \
+    --one-time > "$DIR"/log."$TESTCASE" 2>&1
+assert_link_exists "$ROOT"/link
+assert_file_exists "$ROOT"/link/file
+assert_file_eq "$ROOT"/link/file "$TESTCASE"
+pass
+
+# Test default syncing
+testcase "default-sync"
+echo "$TESTCASE 1" > "$REPO"/file
+git -C "$REPO" commit -qam "$TESTCASE 1"
+$GIT_SYNC \
+    --repo="$REPO" \
+    --root="$ROOT" \
+    --dest="link" > "$DIR"/log."$TESTCASE" 2>&1 &
+sleep 2
+assert_link_exists "$ROOT"/link
+assert_file_exists "$ROOT"/link/file
+assert_file_eq "$ROOT"/link/file "$TESTCASE 1"
+echo "$TESTCASE 2" > "$REPO"/file
+git -C "$REPO" commit -qam "$TESTCASE 2"
+sleep 2
+assert_link_exists "$ROOT"/link
+assert_file_exists "$ROOT"/link/file
+assert_file_eq "$ROOT"/link/file "$TESTCASE 2"
+git -C "$REPO" reset -q --hard HEAD^
+sleep 2
+assert_link_exists "$ROOT"/link
+assert_file_exists "$ROOT"/link/file
+assert_file_eq "$ROOT"/link/file "$TESTCASE 1"
+kill %1
+pass
+
+# Test HEAD syncing
+testcase "head-sync"
+echo "$TESTCASE 1" > "$REPO"/file
+git -C "$REPO" commit -qam "$TESTCASE 1"
+$GIT_SYNC \
+    --repo="$REPO" \
+    --branch=master \
+    --rev=HEAD \
+    --root="$ROOT" \
+    --dest="link" > "$DIR"/log."$TESTCASE" 2>&1 &
+sleep 2
+assert_link_exists "$ROOT"/link
+assert_file_exists "$ROOT"/link/file
+assert_file_eq "$ROOT"/link/file "$TESTCASE 1"
+echo "$TESTCASE 2" > "$REPO"/file
+git -C "$REPO" commit -qam "$TESTCASE 2"
+sleep 2
+assert_link_exists "$ROOT"/link
+assert_file_exists "$ROOT"/link/file
+assert_file_eq "$ROOT"/link/file "$TESTCASE 2"
+git -C "$REPO" reset -q --hard HEAD^
+sleep 2
+assert_link_exists "$ROOT"/link
+assert_file_exists "$ROOT"/link/file
+assert_file_eq "$ROOT"/link/file "$TESTCASE 1"
+kill %1
+pass
+
+# Test branch syncing
+testcase "branch-sync"
+git -C "$REPO" checkout -q -b mybranch
+echo "$TESTCASE 1" > "$REPO"/file
+git -C "$REPO" commit -qam "$TESTCASE 1"
+$GIT_SYNC \
+    --repo="$REPO" \
+    --branch=mybranch \
+    --root="$ROOT" \
+    --dest="link" > "$DIR"/log."$TESTCASE" 2>&1 &
+sleep 2
+assert_link_exists "$ROOT"/link
+assert_file_exists "$ROOT"/link/file
+assert_file_eq "$ROOT"/link/file "$TESTCASE 1"
```

Comment:
I think the issue doesn't come up when the latest revision is on the pulled branch, but when a new rev is pushed to another branch. Therefore adding the commit to the master branch should update the repo, even so it shouldn't be updated as the commit is from another branch. The test fails with the change. Please correct me, if I misinterpreted the test.